src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,609 @@
+        case Arch.ARM64:
+            return "arm64"
+        case _:
+            raise UnsupportedArchitectureError(arch)
+
+
+IMAGE_MOUNT_DIR = Path("/mnt/ubuntu-image/")
+NETWORK_BLOCK_DEVICE_PATH = Path("/dev/nbd0")
+NETWORK_BLOCK_DEVICE_PARTITION_PATH = Path("/dev/nbd0p1")
+
+
+def _clean_build_state() -> None:
+    """Remove any artefacts left by previous build."""
+    # The commands will fail if artefacts do not exist and hence there is no need to check the
+    # output of subprocess runs.
+    IMAGE_MOUNT_DIR.mkdir(parents=True, exist_ok=True)
+    subprocess.run(
+        ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "dev")], timeout=30, check=False
+    )  # nosec: B603
+    subprocess.run(
+        ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "proc")], timeout=30, check=False
+    )  # nosec: B603
+    subprocess.run(
+        ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "sys")], timeout=30, check=False
+    )  # nosec: B603
+    subprocess.run(
+        ["/usr/bin/umount", str(IMAGE_MOUNT_DIR)], timeout=30, check=False
+    )  # nosec: B603
+    subprocess.run(
+        ["/usr/bin/umount", str(NETWORK_BLOCK_DEVICE_PATH)], timeout=30, check=False
+    )  # nosec: B603
+    subprocess.run(  # nosec: B603
+        ["/usr/bin/umount", str(NETWORK_BLOCK_DEVICE_PARTITION_PATH)], timeout=30, check=False
+    )
+    subprocess.run(  # nosec: B603
+        ["/usr/bin/qemu-nbd", "--disconnect", str(NETWORK_BLOCK_DEVICE_PATH)],
+        timeout=30,
+        check=False,
+    )
+    subprocess.run(  # nosec: B603
+        ["/usr/bin/qemu-nbd", "--disconnect", str(NETWORK_BLOCK_DEVICE_PARTITION_PATH)],
+        timeout=30,
+        check=False,
+    )
+
+
+CLOUD_IMAGE_URL_TMPL = (
+    "https://cloud-images.ubuntu.com/{BASE_IMAGE}/current/"
+    "{BASE_IMAGE}-server-cloudimg-{BIN_ARCH}.img"
+)
+CLOUD_IMAGE_FILE_NAME = "{BASE_IMAGE}-server-cloudimg-{BIN_ARCH}.img"
+
+
+class CloudImageDownloadError(Exception):
+    """Represents an error downloading cloud image."""
+
+
+def _download_cloud_image(arch: Arch, base_image: BaseImage) -> Path:
+    """Download the cloud image from cloud-images.ubuntu.com.
+
+    Args:
+        arch: The cloud image architecture to download.
+        base_image: The ubuntu base image OS to download.
+
+    Returns:
+        The downloaded cloud image path.
+
+    Raises:
+        CloudImageDownloadError: If there was an error downloading the image.
+    """
+    try:
+        bin_arch = _get_supported_runner_arch(arch)
+    except UnsupportedArchitectureError as exc:
+        raise CloudImageDownloadError from exc
+
+    try:
+        # The ubuntu-cloud-images is a trusted source
+        image_path, _ = urllib.request.urlretrieve(  # nosec: B310
+            CLOUD_IMAGE_URL_TMPL.format(BASE_IMAGE=base_image.value, BIN_ARCH=bin_arch),
+            CLOUD_IMAGE_FILE_NAME.format(BASE_IMAGE=base_image.value, BIN_ARCH=bin_arch),
+        )
+        return Path(image_path)
+    except urllib.error.ContentTooShortError as exc:
```

Comment:
Do you think there are other errors (e.g. http 5xx) that should be caught?

Reply:
I think 5xx errors should bubble up and let the application throw a stack trace - the only error I could foresee getting raised from the LXD codebase was this error :)

Reply:
I would suggest also catching other urllib/network errors that are possible, as the stack trace is populated to the user without any further explanation, which may not be user friendly.

Reply:
Updated, thank you!